### PR TITLE
Migrate to openapi-generator and okhttp3

### LINKED ISSUE #482: Use of "async" calls eventually causes StackoverflowErrors due to unbounded growth of networkInterceptors in OkHttpClient
I mentioned this in #467 but I thought it was worth opening a new issue for visibility.

Using the generated "async" methods, such as `CoreV1Api.listPodForAllNamespacesAsync(..)`, with an ApiCallback will eventually cause a StackoverflowError because the generated client code adds a new `networkInteceptor` onto the ApiClient's list of networkInterceptors _for every single call_ and never cleans up the list.

For example, `listPodForAllNamespacesAsync` creates a ProgressListener when `callback != null`. When the call stack reaches `listPodForAllNamespacesCall`, this series of generated code does:

```java
if(progressListener != null) {
    apiClient.getHttpClient().networkInterceptors().add(new com.squareup.okhttp.Interceptor() {
      ...
    });
}
```

`networkInterceptors()` is a plain `java.util.List` which grows unbounded in size. The "interceptor chain" concept in OkHttpClient recurses through all of the interceptors in the list, and if the list of interceptors becomes long enough it will eventually cause a StackOverflowError:

```
 Uncaught exception in thread Thread[OkHttp Dispatcher,5,main]
 	java.lang.StackOverflowError: null
 	at io.kubernetes.client.apis.CoreV1Api$581.intercept(CoreV1Api.java:18426)
 	at com.squareup.okhttp.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:695)
 	at io.kubernetes.client.apis.CoreV1Api$581.intercept(CoreV1Api.java:18426)
 	at com.squareup.okhttp.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:695)
 	at io.kubernetes.client.apis.CoreV1Api$581.intercept(CoreV1Api.java:18426)
 	at com.squareup.okhttp.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:695)
 	at io.kubernetes.client.apis.CoreV1Api$581.intercept(CoreV1Api.java:18426)
 	at com.squareup.okhttp.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:695)
 	at io.kubernetes.client.apis.CoreV1Api$581.intercept(CoreV1Api.java:18426)
 	at com.squareup.okhttp.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:695)
 	at io.kubernetes.client.apis.CoreV1Api$581.intercept(CoreV1Api.java:18426)
 	at com.squareup.okhttp.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:695)
...
```

The upstream issue for the swagger-codegen is https://github.com/swagger-api/swagger-codegen/issues/7876 and a fix is proposed in https://github.com/swagger-api/swagger-codegen/pull/8053 but has received no attention for 8+ months.

As-is (along with the thread safety issues in #467), the "async" methods in the generated client code here are virtually unusable in any sort of long-lived application or production server.

### PR DESCRIPTION
**Pull request by fabiokung**
Replaces #595. Requires yue9944882/gen#1 and sundrio/sundrio#156, ~`kubernetes/pom.xml` needs to be updated in this PR after it gets merged~.

Fixes #482.

@yue9944882 @brendanburns I got everything compiling with these changes. They will break backwards compatibility, so a new major version release will be required.

EDIT: more requirements: OpenAPITools/openapi-generator#4226, OpenAPITools/openapi-generator#4182

### CODE REVIEW

**Comment by fabiokung** [comment]
Tests are also passing with a patched `sundrio`. The only way I can think ok making CI happy before sundrio/sundrio/156 gets merged and released would be to vendor a patched jar in here (ugh).

**Comment by fabiokung** [comment]
I take it back, bootstrapping SSL needs some work (okhttp3 does not like a `null` `SSLContextFactory`), and the YAML test needs some fixing since the schema changed for `IntOrString`.

**Comment by brendandburns** [comment]
What do you mean "Schema changed for IntOrString" the Schema should definitely _not_ change in the API.

The last time I tried this, the custom type registration stuff in the generator (e.g. the stuff that makes IntOrString work correctly) was broken, that was the main problem that prevented us from switching.

Please make sure that IntOrString, Quantity, and the other custom types are parsing correctly.

Thanks!

**Comment by fabiokung** [comment]
I see. I meant that I was just assuming the new generator explicitly wanted to map IntOrString to String. 

I missed the custom generators stuff. Your comment makes sense. I will take a look later today or tomorrow.

**Comment by brendandburns** [comment]
I had a thread (a while ago) with one of the OpenAPI maintainers. If you send mail to bburns [at] microsoft I will try to add you to the thread where I discussed some of the challenges I saw...

**Comment by yue9944882** [comment]
hi @fabiokung any plans to revive this thread?

**Comment by fabiokung** [comment]
Hi! Yes. I just had no time in the past weeks, but I can spend some cycles early next week, unless you want to pick it up?

**Comment by yue9944882** [comment]
appreciated if you can continue the work! i added this to my list, and will be coming back for review soon!

**Comment by yue9944882** [comment]
> [ERROR] Failed to execute goal on project client-java-api: Could not resolve dependencies for project io.kubernetes:client-java-api:bundle:7.0.0-SNAPSHOT: Failure to find io.sundr:builder-annotations:jar:0.19-SNAPSHOT in https://oss.sonatype.org/content/repositories/snapshots/ was cached in the local repository, resolution will not be reattempted until the update interval of sonatype-snapshots has elapsed or updates are forced -> [Help 1]

CI's complaining about failing to find `io.sundr:builder-annotations:jar:0.19-SNAPSHOT`, guess we should probably drop the suffix?

**Comment by yue9944882** [comment]
@fabiokung the diff is so large that my browser can hardly load it, i did review locally in my IDE:
1. pls generate the project from the latest release-1.15 branch, we are not skipping 1.15.

2. plz add the following content back to the `.openapi-generator-ignore` file
```
# Remove this once swagger-codegen 2.3.0 is released and we update. 
# Verify the following PRs are in the release:
#   * https://github.com/swagger-api/swagger-codegen/pull/6230
src/main/java/io/kubernetes/client/JSON.java
# src/main/java/io/kubernetes/client/ApiClient.java (NOTE: probably want to add this back for further manual edits)
```

3. plz resolve the git conflicts

4. i see unexpected diffs in `util/src/main/java/io/kubernetes/client/informer/SharedInformerFactory.java`, split this out as another PR:

```
    if (apiClient.getHttpClient().readTimeoutMillis() > 0) {
      log.warn("HttpClient has a read timeout, removing it so watch calls don't terminate early");
      OkHttpClient httpClient =
          apiClient.getHttpClient().newBuilder().readTimeout(0, TimeUnit.MILLISECONDS).build();
      apiClient.setHttpClient(httpClient);
    }
```

5. in `util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java` line 109, why is the try-catch block dropped?

--- Thanks, Min

**Comment by yue9944882** [comment]
@brendandburns do you think we should cut an alpha release for this pull? it will be breaking compatiblities and we can probably missing sth in this migration.. i can probably start by trying the alpha release out in our internal projects.

**Comment by fabiokung** [comment]
> CI's complaining about failing to find io.sundr:builder-annotations:jar:0.19-SNAPSHOT, guess we should probably drop the suffix?

I mentioned this in the PR description, we need to wait until sundrio/sundrio#156 gets merged, then update this to the right released version. In order to test this I had a locally built snapshot in my local maven repo, with the patched sundrio.

**Comment by fabiokung** [comment]
@yue9944882 comments inline:

> 1. pls generate the project from the latest release-1.15 branch, we are not skipping 1.15.

Sure, let me give it a try.

> 2. plz add the following content back to the `.openapi-generator-ignore` file
> 
> ```
> # Remove this once swagger-codegen 2.3.0 is released and we update. 
> # Verify the following PRs are in the release:
> #   * https://github.com/swagger-api/swagger-codegen/pull/6230
> src/main/java/io/kubernetes/client/JSON.java
> # src/main/java/io/kubernetes/client/ApiClient.java (NOTE: probably want to add this back for further manual edits)
> ```

I'm reviewing the custom types with the new openapi-generator, and we may not need this anymore with openapi-generator.

> 
> 1. plz resolve the git conflicts
> 2. i see unexpected diffs in `util/src/main/java/io/kubernetes/client/informer/SharedInformerFactory.java`, split this out as another PR:

These are not unrelated changes, they are part of moving to okhttp3 (used by openapi-generator). The previous swagger-codegen used an older version of okhttp.

> 
> 1. in `util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java` line 109, why is the try-catch block dropped?

Same thing, the `close()` method with okhttp3 does not throw an `IOException` anymore.

**Comment by fabiokung** [comment]
Regenerated against 1.15, and fixed the merge conflict.

Now we wait for OpenAPITools/openapi-generator#4182 to get `IntOrString` and `Quantity` working properly.

**Comment by fabiokung** [comment]
Custom types look OK now. The last remaining issue is that SSL bootstraping in ApiClient (as generated by openapi-generator) is currently throwing `NullPointerException`.

**Comment by brendandburns** [comment]
Thanks for the continued effort here, I'm looking forward to merging this!

**Comment by fabiokung** [comment]
@yue9944882 @brendanburns All green on CI 🎉. PTAL

File: kubernetes/.openapi-generator-ignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+.gitignore
+.travis.yml
+*gradle*
+*sbt*
+git_push.sh
+README.md
```
[the remark this answers is not in the document]
Comment (fabiokung) [reply kubernetes/.openapi-generator-ignore]: Done. I will unpin `ApiClient`, but reapply the date formatter changes and add `JSON.java` to the ignore file.

**Comment by fabiokung** [comment]
The CI failure is a bit odd, it's referencing a line of code that doesn't exist, and I can't reproduce it (https://travis-ci.org/kubernetes-client/java/builds/601953560#L6041). CI flakiness?

**Comment by yue9944882** [comment]
> The CI failure is a bit odd, it's referencing a line of code that doesn't exist, and I can't reproduce it (https://travis-ci.org/kubernetes-client/java/builds/601953560#L6041). CI flakiness?

it's not flakiness. rebasing the pull onto the latest master branch will solve it.

**Comment by fabiokung** [comment]
@yue9944882 rebasing against master didn't seem to fix the CI issue.

**Comment by fabiokung** [comment]
@yue9944882 AFAICT all the necessary changes have been merged to `openapi-generator`, and `sundrio`. I will leave it up to you if pointing kubernetes-client/gen to the `master` branch of `openapi-generator` is acceptable and regenerating all clients and models with it, or if we need to wait for a formal release of `openapi-generator` first.

**Comment by brendanburns** [comment]
Master branch is fine with me. Thanks for the effort here!

**Comment by yue9944882** [comment]
had confirmed w/ @wing328, the next release 4.2.0 will be landing by the end of october. https://github.com/OpenAPITools/openapi-generator/milestone/25. how about wait for 5-ish days? @brendandburns @fabiokung

**Comment by brendanburns** [comment]
Waiting 5 days is fine with me. Thanks for the update.

**Comment by fabiokung** [comment]
> how about wait for 5-ish days?

sgtm, no worries

**Comment by yue9944882** [comment]
@fabiokung can you rename the generated package name to `io.kubernetes.swagger` per #744？ the old project structure is causing issues in OSGi integration. besides i think we should reserve the old sources and deprecate them by following https://stackoverflow.com/a/23019294.

```java
/**
 * @deprecated As of release 7.0.0, replaced by {@link io.kubernetes.swagger;}
 */
@Deprecated
package io.kubernetes.client; // TODO: completely remove it after release 9.0.0
```

**Comment by fabiokung** [comment]
I will give that a try, not sure some of the non-generated classes will be backwards compatible, so keeping old classes may not do much. I will update with more info as I learn more.

**Comment by yue9944882** [comment]
> not sure some of the non-generated classes will be backwards compatible,

can you elaborate? there's no non-generated classes under `client-java-api` module AFAIK. i agree it won't save us from breaking compatibility but at least comfort those users who find their project broken from compiling after upgrading to new libraries in some sense..

**Comment by yue9944882** [comment]
@fabiokung i changed my mind, let's completely remove the old classes and re-generated w/ the new name `io.kubernetes.openapi`. it's okay to lost compatibility between major versions.

**Comment by fabiokung** [comment]
Awesome! I've been pretty swamped this week, but will get to it asap.

**Comment by yue9944882** [comment]
@fabiokung anything i can help to move it forward, thanks?

**Comment by fabiokung** [comment]
On my end I still haven't been able to carve up some time to work on it. I expect to be able to get it done by the end of this week. Would that work for you? If not, we can find ways to hand over the remaining work to you.

**Comment by yue9944882** [comment]
@fabiokung @brendandburns i just opened a pull to continue the work from this pull https://github.com/kubernetes-client/java/pull/778, PTAL

**Comment by fabiokung** [comment]
@yue9944882 would you rather I bring your commits to this PR and rebase everything here, or continue on your PR? There's some important history here, but either way works for me.

**Comment by yue9944882** [comment]
it's fine to separate the discussion/history and the actual pull #778 i suppose. thank you for your great effort in moving this forward!

**Comment by fabiokung** [comment]
continuing in #778

**Pull request closed without merging**